[Flaky Test]: TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged) – error installing package: failed to copy source directory... /opt/Elastic/Agent/elastic-agent: file exists

### Failing test case

TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)

### Error message

/opt/Elastic/Agent/elastic-agent: file exists during install, looks like a previous install was not cleaned or is still being cleaned

### Build

https://buildkite.com/elastic/elastic-agent/builds/6855#018d7f0a-a775-42a3-aa0b-0fb496da1a97

### OS

Linux

### Stacktrace and notes

```markdown
fixture.go:631: >> running binary with: [/tmp/TestStandaloneUpgradeUpgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)332107157/001/elastic-agent-8.12.0-SNAPSHOT-linux-arm64/elastic-agent install --force --non-interactive]
    upgrade_standalone_test.go:77: 
        	Error Trace:	/home/ubuntu/agent/testing/integration/upgrade_standalone_test.go:77
        	            				/home/ubuntu/agent/testing/integration/upgrade_standalone_test.go:52
        	Error:      	Received unexpected error:
        	            	failed to install start agent (err: error running agent install command: exit status 1) [output: Installing in non-interactive mode.
        	            	
[    ] Copying install files  [0s] 
                                   

[   =] Error copying files  [12s] Error: error installing package: failed to copy source directory (/tmp/TestStandaloneUpgradeUpgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)332107157/001/elastic-agent-8.12.0-SNAPSHOT-linux-arm64) to destination (/opt/Elastic/Agent): symlink data/elastic-agent-bf5180/elastic-agent /opt/Elastic/Agent/elastic-agent: file exists
        	            	For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
        	            	]
        	Test:       	TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)
--- FAIL: TestStandaloneUpgrade/Upgrade_8.12.0-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged) (56.84s)
```

There are more tests failing in the same build with the same root cause:
- TestStandaloneUpgradeUninstallKillWatcher
- TestStandaloneUpgradeRollback
- TestStandaloneUpgrade/Upgrade_8.12.1-SNAPSHOT_to_8.13.0-SNAPSHOT_(privileged)

Also in the same build on the same runner, test `TestStandaloneUpgradeRetryDownload` fails because the agent disappears mid-test:

```
  upgrader.go:363: waiting for upgrade watcher to finish
    upgrader.go:368: upgrade watcher finished
    upgrader.go:373: Checking upgrade details state after successful upgrade
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent version --binary-only --yaml]
    upgrade_standalone_retry_test.go:111: 
        	Error Trace:	/home/ubuntu/agent/testing/integration/upgrade_standalone_retry_test.go:111
        	Error:      	Received unexpected error:
        	            	failed to get Agent version: fork/exec /opt/Elastic/Agent/elastic-agent: no such file or directory
        	Test:       	TestStandaloneUpgradeRetryDownload
    fixture_install.go:165: [test TestStandaloneUpgradeRetryDownload] Inside fixture cleanup function
    fixture_install.go:180: collecting diagnostics; test failed
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent diagnostics -f /home/ubuntu/agent/build/diagnostics/TestStandaloneUpgradeRetryDownload-diagnostics-2024-02-06T16:11:30Z.zip]
    fixture_install.go:306: failed to collect diagnostics to /home/ubuntu/agent/build/diagnostics/TestStandaloneUpgradeRetryDownload-diagnostics-2024-02-06T16:11:30Z.zip (fork/exec /opt/Elastic/Agent/elastic-agent: no such file or directory): 
    fixture_install.go:319: creating zip archive of the installation directory: /opt/Elastic/Agent
    fixture_install.go:349: failed to add /opt/Elastic/Agent/elastic-agent to zip, continuing: open /opt/Elastic/Agent/elastic-agent: no such file or directory
    fixture_install.go:349: failed to add /opt/Elastic/Agent/elastic-agent.sock to zip, continuing: open /opt/Elastic/Agent/elastic-agent.sock: no such device or address
    fixture.go:631: >> running binary with: [/opt/Elastic/Agent/elastic-agent uninstall --force]
    fixture_install.go:209: fixture.Install Cleanup: agent was already uninstalled, skipping uninstall
```

This seems like a slow watcher or another slow test cleaning up...

It's curious as it only seem to happen on the arm64 runner
```

--- a/internal/pkg/agent/application/upgrade/rollback.go
+++ b/internal/pkg/agent/application/upgrade/rollback.go
@@ -8,6 +8,7 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"strings"
 	"time"
@@ -138,10 +139,10 @@ func Cleanup(log *logger.Logger, topDirPath, currentVersionedHome, currentHash s
 
 // InvokeWatcher invokes an agent instance using watcher argument for watching behavior of
 // agent during upgrade period.
-func InvokeWatcher(log *logger.Logger, agentExecutable string) error {
+func InvokeWatcher(log *logger.Logger, agentExecutable string) (*exec.Cmd, error) {
 	if !IsUpgradeable() {
 		log.Info("agent is not upgradable, not starting watcher")
-		return nil
+		return nil, nil
 	}
 
 	cmd := invokeCmd(agentExecutable)
@@ -154,14 +155,14 @@ func InvokeWatcher(log *logger.Logger, agentExecutable string) error {
 
 	log.Infow("Starting upgrade watcher", "path", cmd.Path, "args", cmd.Args, "env", cmd.Env, "dir", cmd.Dir)
 	if err := cmd.Start(); err != nil {
-		return fmt.Errorf("failed to start Upgrade Watcher: %w", err)
+		return nil, fmt.Errorf("failed to start Upgrade Watcher: %w", err)
 	}
 
 	upgradeWatcherPID := cmd.Process.Pid
 	agentPID := os.Getpid()
 	log.Infow("Upgrade Watcher invoked", "agent.upgrade.watcher.process.pid", upgradeWatcherPID, "agent.process.pid", agentPID)
 
-	return nil
+	return cmd, nil
 }
 
 func restartAgent(ctx context.Context, log *logger.Logger, c client.Client) error {

--- a/internal/pkg/agent/application/upgrade/upgrade.go
+++ b/internal/pkg/agent/application/upgrade/upgrade.go
@@ -10,9 +10,11 @@ import (
 	"fmt"
 	"io/fs"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"runtime"
 	"strings"
+	"time"
 
 	"github.com/otiai10/copy"
 	"go.elastic.co/apm"
@@ -49,6 +51,8 @@ var agentArtifact = artifact.Artifact{
 	Cmd:      agentName,
 	Artifact: "beats/" + agentName,
 }
+
+var ErrWatcherNotStarted = errors.New("watcher did not start in time")
 
 // Upgrader performs an upgrade
 type Upgrader struct {
@@ -297,10 +301,18 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 		// use the new agent executable as it should be able to parse the new update marker
 		watcherExecutable = paths.BinaryPath(filepath.Join(paths.Top(), unpackRes.VersionedHome), agentName)
 	}
-	if err := InvokeWatcher(u.log, watcherExecutable); err != nil {
+	var watcherCmd *exec.Cmd
+	if watcherCmd, err = InvokeWatcher(u.log, watcherExecutable); err != nil {
 		u.log.Errorw("Rolling back: starting watcher failed", "error.message", err)
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
-		return nil, err
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(err, rollbackErr)
+	}
+
+	watcherWaitErr := waitForWatcher(ctx, u.log, markerFilePath(paths.Data()), 30*time.Second)
+	if watcherWaitErr != nil {
+		killWatcherErr := watcherCmd.Process.Kill()
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(watcherWaitErr, killWatcherErr, rollbackErr)
 	}
 
 	cb := shutdownCallback(u.log, paths.Home(), release.Version(), version, filepath.Join(paths.Top(), unpackRes.VersionedHome))
@@ -313,6 +325,31 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	}
 
 	return cb, nil
+}
+
+func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath string, waitTime time.Duration) error {
+	//Wait for the watcher to be up and running
+	watcherContext, cancel := context.WithTimeout(ctx, waitTime)
+	defer cancel()
+
+	markerWatcher := newMarkerFileWatcher(markerFilePath, log)
+	err := markerWatcher.Run(watcherContext)
+	if err != nil {
+		return fmt.Errorf("error starting update marker watcher: %w", err)
+	}
+
+	for {
+		select {
+		case updMarker := <-markerWatcher.Watch():
+			if updMarker.Details != nil && updMarker.Details.State == details.StateWatching {
+				// watcher started and it is watching, all good
+				return nil
+			}
+
+		case <-watcherContext.Done():
+			return ErrWatcherNotStarted
+		}
+	}
 }
 
 // Ack acks last upgrade action

--- a/internal/pkg/agent/application/upgrade/upgrade.go
+++ b/internal/pkg/agent/application/upgrade/upgrade.go
@@ -6,6 +6,7 @@ package upgrade
 
 import (
 	"context"
+	goerrors "errors"
 	"fmt"
 	"io/fs"
 	"os"
@@ -257,8 +258,8 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 
 	if err := changeSymlink(u.log, paths.Top(), symlinkPath, newPath); err != nil {
 		u.log.Errorw("Rolling back: changing symlink failed", "error.message", err)
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
-		return nil, err
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(err, rollbackErr)
 	}
 
 	// We rotated the symlink successfully: prepare the current and previous agent installation details for the update marker
@@ -283,8 +284,8 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 		previous,     // old agent version data
 		action, det); err != nil {
 		u.log.Errorw("Rolling back: marking upgrade failed", "error.message", err)
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
-		return nil, err
+		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
+		return nil, goerrors.Join(err, rollbackErr)
 	}
 
 	minParsedVersionForNewUpdateMarker := agtversion.NewParsedSemVer(8, 13, 0, "", "")
@@ -378,18 +379,19 @@ func isSameVersion(log *logger.Logger, current agentVersion, metadata packageMet
 	return current == newVersion, newVersion
 }
 
-func rollbackInstall(ctx context.Context, log *logger.Logger, topDirPath, versionedHome, oldVersionedHome string) {
-	newAgentInstallPath := filepath.Join(topDirPath, versionedHome)
-	err := os.RemoveAll(newAgentInstallPath)
+func rollbackInstall(ctx context.Context, log *logger.Logger, topDirPath, versionedHome, oldVersionedHome string) error {
+	oldAgentPath := paths.BinaryPath(filepath.Join(topDirPath, oldVersionedHome), agentName)
+	err := changeSymlink(log, topDirPath, filepath.Join(topDirPath, agentName), oldAgentPath)
 	if err != nil && !errors.Is(err, fs.ErrNotExist) {
-		log.Warnw(fmt.Sprintf("rolling back install: removing new agent install at %q failed", newAgentInstallPath), "error.message", err)
+		return fmt.Errorf("rolling back install: restoring symlink to %q failed: %w", oldAgentPath, err)
 	}
 
-	oldAgentPath := paths.BinaryPath(filepath.Join(topDirPath, oldVersionedHome), agentName)
-	err = changeSymlink(log, topDirPath, filepath.Join(topDirPath, agentName), oldAgentPath)
+	newAgentInstallPath := filepath.Join(topDirPath, versionedHome)
+	err = os.RemoveAll(newAgentInstallPath)
 	if err != nil && !errors.Is(err, fs.ErrNotExist) {
-		log.Warnw(fmt.Sprintf("rolling back install: restoring symlink to %q failed", oldAgentPath), "error.message", err)
+		return fmt.Errorf("rolling back install: removing new agent install at %q failed: %w", newAgentInstallPath, err)
 	}
+	return nil
 }
 
 func copyActionStore(log *logger.Logger, newHome string) error {

--- a/internal/pkg/agent/application/upgrade/upgrade.go
+++ b/internal/pkg/agent/application/upgrade/upgrade.go
@@ -53,7 +53,7 @@ var agentArtifact = artifact.Artifact{
 	Artifact: "beats/" + agentName,
 }
 
-var ErrWatcherNotStarted = errors.New("watcher did not start in time or context expired")
+var ErrWatcherNotStarted = errors.New("watcher did not start in time")
 
 // Upgrader performs an upgrade
 type Upgrader struct {
@@ -329,7 +329,7 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 }
 
 func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath string, waitTime time.Duration) error {
-	//Wait for the watcher to be up and running
+	// Wait for the watcher to be up and running
 	watcherContext, cancel := context.WithTimeout(ctx, waitTime)
 	defer cancel()
 
@@ -352,7 +352,7 @@ func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath stri
 
 		case <-watcherContext.Done():
 			log.Error("upgrade watcher did not start watching within %s or context has expired", waitTime)
-			return ErrWatcherNotStarted
+			return goerrors.Join(ErrWatcherNotStarted, watcherContext.Err())
 		}
 	}
 }

--- a/internal/pkg/agent/application/upgrade/upgrade.go
+++ b/internal/pkg/agent/application/upgrade/upgrade.go
@@ -39,11 +39,12 @@ import (
 )
 
 const (
-	agentName       = "elastic-agent"
-	hashLen         = 6
-	agentCommitFile = ".elastic-agent.active.commit"
-	runDirMod       = 0770
-	snapshotSuffix  = "-SNAPSHOT"
+	agentName          = "elastic-agent"
+	hashLen            = 6
+	agentCommitFile    = ".elastic-agent.active.commit"
+	runDirMod          = 0770
+	snapshotSuffix     = "-SNAPSHOT"
+	watcherMaxWaitTime = 30 * time.Second
 )
 
 var agentArtifact = artifact.Artifact{
@@ -52,7 +53,7 @@ var agentArtifact = artifact.Artifact{
 	Artifact: "beats/" + agentName,
 }
 
-var ErrWatcherNotStarted = errors.New("watcher did not start in time")
+var ErrWatcherNotStarted = errors.New(fmt.Sprintf("watcher did not start within %s or context expired", watcherMaxWaitTime))
 
 // Upgrader performs an upgrade
 type Upgrader struct {
@@ -308,7 +309,7 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 		return nil, goerrors.Join(err, rollbackErr)
 	}
 
-	watcherWaitErr := waitForWatcher(ctx, u.log, markerFilePath(paths.Data()), 30*time.Second)
+	watcherWaitErr := waitForWatcher(ctx, u.log, markerFilePath(paths.Data()), watcherMaxWaitTime)
 	if watcherWaitErr != nil {
 		killWatcherErr := watcherCmd.Process.Kill()
 		rollbackErr := rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
@@ -338,15 +339,19 @@ func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath stri
 		return fmt.Errorf("error starting update marker watcher: %w", err)
 	}
 
+	log.Info("waiting for upgrade watcher to set %s state in upgrade marker", details.StateWatching)
+
 	for {
 		select {
 		case updMarker := <-markerWatcher.Watch():
 			if updMarker.Details != nil && updMarker.Details.State == details.StateWatching {
 				// watcher started and it is watching, all good
+				log.Info("upgrade watcher set %s state in upgrade marker: exiting wait loop", details.StateWatching)
 				return nil
 			}
 
 		case <-watcherContext.Done():
+			log.Error("upgrade watcher did not start watching within %s or context has expired", waitTime)
 			return ErrWatcherNotStarted
 		}
 	}

--- a/internal/pkg/agent/application/upgrade/upgrade.go
+++ b/internal/pkg/agent/application/upgrade/upgrade.go
@@ -53,7 +53,7 @@ var agentArtifact = artifact.Artifact{
 	Artifact: "beats/" + agentName,
 }
 
-var ErrWatcherNotStarted = errors.New(fmt.Sprintf("watcher did not start within %s or context expired", watcherMaxWaitTime))
+var ErrWatcherNotStarted = errors.New("watcher did not start in time or context expired")
 
 // Upgrader performs an upgrade
 type Upgrader struct {
@@ -339,7 +339,7 @@ func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath stri
 		return fmt.Errorf("error starting update marker watcher: %w", err)
 	}
 
-	log.Info("waiting for upgrade watcher to set %s state in upgrade marker", details.StateWatching)
+	log.Info("waiting up to %s for upgrade watcher to set %s state in upgrade marker", waitTime, details.StateWatching)
 
 	for {
 		select {

--- a/internal/pkg/agent/application/upgrade/upgrade_test.go
+++ b/internal/pkg/agent/application/upgrade/upgrade_test.go
@@ -759,14 +759,14 @@ func TestWaitForWatcher(t *testing.T) {
 			name:                "Happy path: watcher is watching already",
 			states:              []details.State{details.StateWatching},
 			stateChangeInterval: 1 * time.Millisecond,
-			timeout:             5 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
 			wantErr:             assert.NoError,
 		},
 		{
 			name:                "Sad path: watcher is never starting",
 			states:              []details.State{details.StateReplacing},
 			stateChangeInterval: 1 * time.Millisecond,
-			timeout:             5 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
 			wantErr:             wantErrWatcherNotStarted,
 		},
 		{
@@ -788,7 +788,7 @@ func TestWaitForWatcher(t *testing.T) {
 			name:                "Timeout: marker is never created",
 			states:              nil,
 			stateChangeInterval: 1 * time.Millisecond,
-			timeout:             5 * time.Millisecond,
+			timeout:             50 * time.Millisecond,
 			wantErr:             wantErrWatcherNotStarted,
 		},
 		{

--- a/internal/pkg/agent/application/upgrade/upgrade_test.go
+++ b/internal/pkg/agent/application/upgrade/upgrade_test.go
@@ -781,7 +781,7 @@ func TestWaitForWatcher(t *testing.T) {
 				details.StateWatching,
 			},
 			stateChangeInterval: 1 * time.Millisecond,
-			timeout:             100 * time.Millisecond,
+			timeout:             500 * time.Millisecond,
 			wantErr:             assert.NoError,
 		},
 		{

--- a/internal/pkg/agent/application/upgrade/upgrade_test.go
+++ b/internal/pkg/agent/application/upgrade/upgrade_test.go
@@ -11,6 +11,7 @@ import (
 	"os"
 	"path/filepath"
 	"runtime"
+	"sync"
 	"testing"
 	"time"
 
@@ -825,11 +826,17 @@ func TestWaitForWatcher(t *testing.T) {
 				writeState(t, updMarkerFilePath, initialState)
 			}
 
+			wg := new(sync.WaitGroup)
+
 			var furtherStates []details.State
 			if len(tt.states) > 1 {
 				// we have more states to produce
 				furtherStates = tt.states[1:]
+
+				wg.Add(1)
+
 				go func() {
+					defer wg.Done()
 					tick := time.NewTicker(tt.stateChangeInterval)
 					defer tick.Stop()
 					for _, state := range furtherStates {
@@ -847,6 +854,12 @@ func TestWaitForWatcher(t *testing.T) {
 			log, _ := logger.NewTesting(tt.name)
 
 			tt.wantErr(t, waitForWatcher(ctx, log, updMarkerFilePath, tt.timeout), fmt.Sprintf("waitForWatcher %s, %v, %s, %s)", updMarkerFilePath, tt.states, tt.stateChangeInterval, tt.timeout))
+
+			// cancel context
+			cancel()
+
+			//wait for goroutines to finish
+			wg.Wait()
 		})
 	}
 }

--- a/internal/pkg/agent/application/upgrade/upgrade_test.go
+++ b/internal/pkg/agent/application/upgrade/upgrade_test.go
@@ -858,7 +858,7 @@ func TestWaitForWatcher(t *testing.T) {
 			// cancel context
 			cancel()
 
-			//wait for goroutines to finish
+			// wait for goroutines to finish
 			wg.Wait()
 		})
 	}
@@ -884,7 +884,8 @@ func writeState(t *testing.T, path string, state details.State) {
 	})
 
 	bytes, err := yaml.Marshal(ms)
-	require.NoError(t, err, "error marshaling the test upgrade marker")
-	err = os.WriteFile(path, bytes, 0770)
-	require.NoError(t, err, "error writing out the test upgrade marker")
+	if assert.NoError(t, err, "error marshaling the test upgrade marker") {
+		err = os.WriteFile(path, bytes, 0770)
+		assert.NoError(t, err, "error writing out the test upgrade marker")
+	}
 }

--- a/internal/pkg/agent/application/upgrade/upgrade_test.go
+++ b/internal/pkg/agent/application/upgrade/upgrade_test.go
@@ -17,10 +17,12 @@ import (
 	"github.com/gofrs/flock"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
 
 	"github.com/elastic/elastic-agent-libs/transport/httpcommon"
 	"github.com/elastic/elastic-agent-libs/transport/tlscommon"
 	"github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/artifact"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details"
 	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
 	"github.com/elastic/elastic-agent/internal/pkg/config"
 	"github.com/elastic/elastic-agent/internal/pkg/release"
@@ -739,4 +741,137 @@ func TestIsSameVersion(t *testing.T) {
 				log, test.args.current, test.args.metadata, test.args.version, test.want.newVersion)
 		})
 	}
+}
+
+func TestWaitForWatcher(t *testing.T) {
+	wantErrWatcherNotStarted := func(t assert.TestingT, err error, i ...interface{}) bool {
+		return assert.ErrorIs(t, err, ErrWatcherNotStarted, i)
+	}
+	tests := []struct {
+		name                string
+		states              []details.State
+		stateChangeInterval time.Duration
+		timeout             time.Duration
+		wantErr             assert.ErrorAssertionFunc
+	}{
+		{
+			name:                "Happy path: watcher is watching already",
+			states:              []details.State{details.StateWatching},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             5 * time.Millisecond,
+			wantErr:             assert.NoError,
+		},
+		{
+			name:                "Sad path: watcher is never starting",
+			states:              []details.State{details.StateReplacing},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             5 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+		{
+			name: "Runaround path: marker is jumping around and landing on watching",
+			states: []details.State{
+				details.StateRequested,
+				details.StateScheduled,
+				details.StateDownloading,
+				details.StateExtracting,
+				details.StateReplacing,
+				details.StateRestarting,
+				details.StateWatching,
+			},
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             100 * time.Millisecond,
+			wantErr:             assert.NoError,
+		},
+		{
+			name:                "Timeout: marker is never created",
+			states:              nil,
+			stateChangeInterval: 1 * time.Millisecond,
+			timeout:             5 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+		{
+			name: "Timeout2: state doesn't get there in time",
+			states: []details.State{
+				details.StateRequested,
+				details.StateScheduled,
+				details.StateDownloading,
+				details.StateExtracting,
+				details.StateReplacing,
+				details.StateRestarting,
+				details.StateWatching,
+			},
+
+			stateChangeInterval: 5 * time.Millisecond,
+			timeout:             20 * time.Millisecond,
+			wantErr:             wantErrWatcherNotStarted,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			deadline, ok := t.Deadline()
+			if !ok {
+				deadline = time.Now().Add(5 * time.Second)
+			}
+			ctx, cancel := context.WithDeadline(context.TODO(), deadline)
+			defer cancel()
+
+			tmpDir := t.TempDir()
+			updMarkerFilePath := filepath.Join(tmpDir, markerFilename)
+
+			if len(tt.states) > 0 {
+				initialState := tt.states[0]
+				writeState(t, updMarkerFilePath, initialState)
+			}
+
+			var furtherStates []details.State
+			if len(tt.states) > 1 {
+				// we have more states to produce
+				furtherStates = tt.states[1:]
+				go func() {
+					tick := time.NewTicker(tt.stateChangeInterval)
+					defer tick.Stop()
+					for _, state := range furtherStates {
+						select {
+						case <-ctx.Done():
+							return
+						case <-tick.C:
+							writeState(t, updMarkerFilePath, state)
+						}
+					}
+
+				}()
+			}
+
+			log, _ := logger.NewTesting(tt.name)
+
+			tt.wantErr(t, waitForWatcher(ctx, log, updMarkerFilePath, tt.timeout), fmt.Sprintf("waitForWatcher %s, %v, %s, %s)", updMarkerFilePath, tt.states, tt.stateChangeInterval, tt.timeout))
+		})
+	}
+}
+
+func writeState(t *testing.T, path string, state details.State) {
+	ms := newMarkerSerializer(&UpdateMarker{
+		Version:           "version",
+		Hash:              "hash",
+		VersionedHome:     "versionedHome",
+		UpdatedOn:         time.Now(),
+		PrevVersion:       "prev_version",
+		PrevHash:          "prev_hash",
+		PrevVersionedHome: "prev_versionedhome",
+		Acked:             false,
+		Action:            nil,
+		Details: &details.Details{
+			TargetVersion: "version",
+			State:         state,
+			ActionID:      "",
+			Metadata:      details.Metadata{},
+		},
+	})
+
+	bytes, err := yaml.Marshal(ms)
+	require.NoError(t, err, "error marshaling the test upgrade marker")
+	err = os.WriteFile(path, bytes, 0770)
+	require.NoError(t, err, "error writing out the test upgrade marker")
 }

--- a/internal/pkg/agent/cmd/run.go
+++ b/internal/pkg/agent/cmd/run.go
@@ -226,7 +226,7 @@ func runElasticAgent(ctx context.Context, cancel context.CancelFunc, override cf
 	}
 
 	// initiate agent watcher
-	if err := upgrade.InvokeWatcher(l, paths.TopBinaryPath()); err != nil {
+	if _, err := upgrade.InvokeWatcher(l, paths.TopBinaryPath()); err != nil {
 		// we should not fail because watcher is not working
 		l.Error(errors.New(err, "failed to invoke rollback watcher"))
 	}

